nunjucks/src/environment.js

```diff
@@ -476,14 +476,15 @@ class Template extends Obj {
     let didError = false;
 
     this.rootRenderFunc(this.env, context, frame, globalRuntime, (err, res) => {
-      if (didError) {
+      // TODO: this is actually a bug in the compiled template (because waterfall
+      // tasks are both not passing errors up the chain of callbacks AND are not
+      // causing a return from the top-most render function). But fixing that
+      // will require a more substantial change to the compiler.
+      if (didError && cb && typeof res !== 'undefined') {
```

Comment:
We just that `res` is undefined, because the second call to the root render callback happens here in `compileRoot`:

https://github.com/mozilla/nunjucks/blob/63c4baf2332db08ce9c170bafe0d3696d1e293bd/nunjucks/src/compiler.js#L1129-L1133

If an error occurs inside an included template, the error will be passed up the callback chain, and the user's callback will get called with the correct error argument. But it won't cause a return from the root compiled template function, so it will still call `cb(null, this.buffer)` at the very end. `this.buffer` is always a string, even if empty, so by confirming that `res` is not undefined we can check specifically for the situation described above.